pkg/clientgocred/cred.go

```diff
@@ -0,0 +1,332 @@
+		writeCacheFile:                   writeCacheFile,
+		useApplicationDefaultCredentials: options.UseApplicationDefaultCredentials,
+	}
+}
+
+// Options struct inputs to PrintCred
+type Options struct {
+	UseApplicationDefaultCredentials bool
+}
+
+// PrintCred prints ExecCredential to stdout to be consumed by kubectl to connect to GKE Clusters
+// {
+//    "kind": "ExecCredential",
+//    "apiVersion": "client.authentication.k8s.io/v1beta1",
+//    "spec": {
+//        "interactive": false
+//    },
+//    "status": {
+//        "expirationTimestamp": "2022-01-27T07:10:46Z",
+//        "token": "ya29.A0ARrda.......0jDi8weH-36jJNru6Ps"
+//    }
+// }
+func PrintCred(options *Options) error {
+	if options == nil {
+		options = &Options{
+			UseApplicationDefaultCredentials: false,
+		}
+	}
+	pc := newPlugin(options)
+
+	ec, err := pc.execCredential()
+	if err != nil {
+		return err
+	}
+
+	ecStr, err := formatToJSON(ec)
+	if err != nil {
+		return err
+	}
+
+	fmt.Print(ecStr)
+	return nil
+}
+
+// execCredential return an object of type ExecCredential which
+// holds a bearer token to authenticate to GKE.
+func (pc *plugin) execCredential() (*clientauthv1b1.ExecCredential, error) {
+	token, expiry, err := pc.accessToken()
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientauthv1b1.ExecCredential{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ExecCredential",
+			APIVersion: "client.authentication.k8s.io/v1beta1",
+		},
+		Status: &clientauthv1b1.ExecCredentialStatus{
+			Token:               token,
+			ExpirationTimestamp: expiry,
+		},
+	}, nil
+}
+
+// accessToken return either the ApplicationDefaultCredentials or the gcloudAccessToken
+func (pc *plugin) accessToken() (string, *metav1.Time, error) {
+	if !pc.useApplicationDefaultCredentials {
+		token, expiry, err := pc.gcloudAccessToken()
+		if err == nil {
+			return token, expiry, nil
+		}
+		// if err is not nil, fall back to returning Application default credentials
+	}
+	return pc.defaultAccessToken()
+}
+
+// gcloudAccessToken returns a cached token if the token is not expired. If the token is
+// expired, it gets a new access token by invoking gcloud command, caches the new token
+// and returns the token.
+func (pc *plugin) gcloudAccessToken() (string, *metav1.Time, error) {
+	if token, expiry, err := pc.getCachedGcloudAccessToken(); err == nil {
+		return token, expiry, nil
+	}
+
+	gc, err := pc.newGcloudConfig()
+	if err != nil {
+		return "", nil, err
+	}
+
+	pc.writeGcloudAccessTokenToCache(gc.Credential.AccessToken, gc.Credential.TokenExpiry)
+
+	return gc.Credential.AccessToken, &metav1.Time{Time: gc.Credential.TokenExpiry}, nil
+}
+
+func (pc *plugin) defaultAccessToken() (string, *metav1.Time, error) {
+	var tok *oauth2.Token
+
+	// Retries (max 5 retries with approx delay 10*ms+jitter setup) help get around occasional network glitches
+	err := retry.OnError(retry.DefaultRetry, func(err error) bool { return true }, func() error {
```

Comment:
According to the documentation, we should use retry.DefaultBackoff (which is exponential).

retry.DefaultRetry is for cases of resource conflict.  If we use it for cases that could be due to network error, we could end up swamping whatever is failing.

(Actually, having read the docs and written this out, I'm now suspicious that there's no valid situation where you want to use retry.DefaultRetry --- any failure to access a Kubernetes resource could be due to a failing component that you don't want to swamp.)

Reply:
Updated to Backoff.. Technically, this is not a congestion issue.. just needed retries.